library/net.c

```diff
@@ -751,6 +751,9 @@ int dnet_process_recv(struct dnet_backend_io *backend, struct dnet_net_state *st
 
 void dnet_state_remove_nolock(struct dnet_net_state *st)
 {
+	if (st->__join_state == DNET_JOIN)
```

Comment:
JOIN state means this node joined others, i.e. this node had connected, sent JOIN request and auth command.
Client never sends this command, i.e. client never joins cluster.
Even more, when server accepts new client it can move it into dht list without JOIN state, i.e. when client sends JOIN command to given node.

What you need is a number of connected states for given **client** node.
This is a number of states in node->storage_state_list. Access to this list is a bit scattered over the code, but it can be easily refactored to single function. Please note list_move* functions operating this list too.
move operation has to increase number of states only if list_empty(&state->storage_state_entry) is true, i.e. given state has not been added yet.